opm/simulators/wells/WellGroupHelpers.cpp

```diff
@@ -673,6 +674,17 @@ updateGroupProductionRates(const Group& group,
         rates[phase] = sumWellPhaseRates(false, group, schedule, wellState, reportStepIdx, phase, /*isInjector*/ false);
     }
     group_state.update_production_rates(group.name(), rates);
+
+    const auto& network = schedule[reportStepIdx].network();
+    if (network.active()) {
+        std::vector<Scalar> network_rates = rates;
+        if (network.needs_instantaneous_rates(schedule, reportStepIdx)) {
+            for (int phase = 0; phase < np; ++phase) {
+                network_rates[phase] = sumWellPhaseRates(false, group, schedule, wellState, reportStepIdx, phase, /*isInjector*/ false, /*network*/ true);
```

Comment:
I can be wrong here. It looks like for the `network_production_rates`, we only the values for all the `leaf_nodes` in its current form. 

Maybe we should only need to do differently for the `leaf_nodes`.  It looks like we will deal with other non-leaf nodes in the function `computeNetworkPressures` based on the rates of the `leaf_nodes`. 

If that sounds correct to you, maybe we can deal with this in the function `computeNetworkPressures` where the `leaf_nodes` information are available?

Reply:
You are right, only the leaf nodes are used in computeNetworkPressures, but this seemed to be the easies way to get those correctly. Also, to be able to report the rates I expect we will need to have all the rates actually used in the network available. 

I may have missed something, but to handle this entirely within the computeNetworkPressures it appears that large parts of the sumWellPhaseRates function would need to be duplicated (essentially everything below the recursive call), so I fail to see how that would be simpler (?)

Reply:
One important thing missing from current `m_network_production_rates` in the current form is that, if a network node is not in the GRUPTREE, it will not be in the `m_network_production_rates`.  Because network nodes and group nodes can be different. 

And also, can we call `sumWellPhaseRates` through `computeNetworkPressures` in some way instead of duplicating them. I could not tell why we can not do that by looking at the code.

Ideally, we might need a function called `updateNetworkProductionRates()`. You do not have to do it unless you are really up to it. We can make a refactoring PR following this PR.

Reply:
You are correct when coming to the parallel communication. I did think about it and my apologizes.  Will respond with more looking into it.

Reply:
No worries, and thanks! I'll await further input then.

Reply:
(without considering NEFAC, then basically, NEFAC is 1.0 for the nodes not in the GRUPTREE)

In the current way, `m_network_production_rates` holds the values for all the group nodes.  `node_inflows` holds a complete version of the rates for all the network nodes. 

When all the group nodes are in the network, for the nodes they both contain, they should have the same values. although calculated through different paths and NEFAC is 1.0. 

For the situations where some group nodes are not in the network, for example, they have different levels of hierarchy, how to calculate the values (using the efficiency factors) I think is something we can only test to find out, it is not totally clear to me about how things should be done. Likely, `m_network_production_rates` can have different values from the values in `node_inflows`, depending on the understanding we use at the end. 

In the current code, `computeNetworkPressures` is going through the network branches, and `updateGroupProductionRates` is going through the group tree.  We do not want to do global communication in the function `computeNetworkPressures`.  Ideally we can borrow the function `updateGroupProductionRates` to get leaf-nodes (only leaf-nodes), and handle the network-way of production rates in `computeNetworkPressures`. 

Now I agree with you that the current way is a simpler way to get things done. And also, I see that with NEFAC and nodes in the GRUPTREE and not in the network probably introduces re-working in these parts. We can get the results correct first and follow with re-factoring in these parts. 

Sorry for the previous misunderstandings.

Reply:
No problem, I can see it is confusing with the member ```m_network_production_rates``` where only leaf rates are used, and will try to make it more clear with proper comments. Glad we can agree to move forward with the current overall approach for now. Thank you.

Reply:
> When all the group nodes are in the network, for the nodes they both contain, they should have the same values. although calculated through different paths and NEFAC is 1.0.

Just for record in case it helps understanding and design or for later usage.

Not totally sure though. With extended network model, we might have different hierarchy layout between nodes, then for the non-leaf nodes, they will have  very different values between `m_network_production_rates` and `node_inflows`. Luckily, the leaf nodes in the `m_network_production_rates`  should still have the correct/desired value.  `node_inflows` will most likely be the one holding the correct/desired values that represent the network structure.